### ISSUE TITLE
Extra parameters are ignored in the client auth flow

### ISSUE THREAD

**Issue by nataliahering** [panva/node-oidc-provider#258]
The [extra parameters](https://github.com/panva/node-oidc-provider/blob/master/docs/configuration.md#extraparams]) are available in the `ctx.oidc.params`, only in the implicit flow. 
It has been observed that they are not being added in the 2-leg auth flow. 

The whitelist that is being created [here](https://github.com/panva/node-oidc-provider/blob/master/lib/shared/chains/client_auth.js) does not include the extra params from the configuration and hence they are getting ignored.

We have the need to pass extra information in the 2-leg flow. Possible solutions:
1) During the first step, store the extra params value in the DB (sessions)
2) Update the `client_auth` and whitelist also the extra params 
`const extras = instance(provider).configuration('extraParams');`
 `extras.forEach(whitelist.add.bind(whitelist));`
  `const buildParams = params(whitelist);`
3) Use the client_secret_jwt auth and pass the extra information as extra claims in the jwt token. This can be decoded and the information can be added to the `ctx.oidc.params`.

The (2) and (3) have the drawback that the extra information can't be passed from the first step.

Any ideas?

**Comment by panva** [comment]
> The extra parameters are available in the ctx.oidc.params, _only in the implicit flow_.

That's incorrect, they're being parsed and available for interactions and helpers such as `interactionCheck` in every `authorization_endpoint` request regardless of the response_type. The original intention was so that e.g. tracking parameters can be passed to interactions for trusted clients and then displayed in the views.

It is working exactly as documented.

```
extraParams
Pass an iterable object (i.e. Array or set of strings) to extend the parameters recognized by the **authorization endpoint**. These parameters are then available in ctx.oidc.params as well as passed to interaction session details

affects: **authorization, interaction**
```

I'm not sure what you're trying to achieve, which grant_type are trying to change/are implementing. Please provide more details.

**Comment by nataliahering** [comment]
Thanks for the prompt reply. Here are some more details.
Assume we have set up the config to include `extraParams: ['extraParameter']`.
Flow:
step 1: `GET` to `oidc/auth` endpoint, `response_type=code`, `extraParameter=extraValue`
step 2: `POST` to `oidc/token` endpoint, using the `client_secret_basic` flow and the code returned from step 1. Include the extraParameter=extraValue in the body.

result: It is correct that the extra parameter is not getting ignored when the server processes the get request in step 1. However, when the post request (step 2) is processed (and that's when we need it), the extra parameter does not exist anywhere in the DB and hence we cannot get it. If we pass it in the body of the post request, it will get ignored. 
When the `findById` is getting called during step 2, we want to have the extra parameter included in the `ctx.oidc.params`.

alternative step 2: `POST` to `oidc/token`, using the `client_secret_jwt` flow. The `client_assertion` jwt token includes the extraParameter=extraValue as an extra claim. 

result: The `client_assertion` token gets decoded and successfully validated, but the extra parameter does not get added in the `oidc.params`.

Hope the flow is more clear now.

**Comment by panva** [comment]
> Hope the flow is more clear now.

It's really not so much :) The call to /token shouldn't need any extra parameters for conformant behavior, but anyway...


> However, when the post request (step 2) is processed (and that's when we need it), the extra parameter does not exist anywhere in the DB and hence we cannot get it.

`extraParams` is only for the authorization_endpoint and does not mean parameters are stored with the or available in the token_endpoint call to you. The only things stored are the ones necessary for the protocol exchange and tokens having the right properties. From my POV this is working as intended, I do not support passing custom parameters down to a token(authorization code in this case) level and having them magically available.

> When the findById is getting called during step 2, we want to have the extra parameter included in the ctx.oidc.params.

If just submitting stuff to /token and having it available in the ctx is what you're after you can access the raw oidc-provider parsed `ctx.oidc.body` or ctx.query

**Comment by nataliahering** [comment]
I see. Would this https://github.com/panva/node-oidc-provider/pull/259 be accepted?
We could then get the extra parameters in the mentioned flow.

**Comment by panva** [comment]
It would not, i'm sorry. why don't you just reach out to `ctx.oidc.body` for the parameter you're expecting?

**Comment by SokratisVidros** [comment]
My two cents, @panva I see your point, however it would be more intuitive for the `extraParams` configuration to apply also to the `/token` endpoint. 

That way, clients implementing the auth code flow, can store the initial params of the `/auth` request in their own session storage mechanism and then piggyback them in the `/token` request during the second leg of the flow.

Hope that helps.

**Comment by panva** [comment]
I feel like you're expecting extraParams to do something it doesn't do anyway. Anything you submit to /auth (and whitelisted as extraParams) won't be available in any other place then the authorization and interaction contexts. That's intended. If extraParams gets applied to /token it **will not** make parameters passthrough from auth to token.

Bending extraParams to do more than it was intended to do is not something i wish to do.

If you have the need to access **any** parameter in the token (client authenticated) endpoints you can access `ctx.oidc.body`. That either solves the issue for you or you still need something that wasn't explained enough.

**Comment by SokratisVidros** [comment]
Yes, the issue is solved with the `ctx.oidc.body` hack. My suggestion is not related to the expectation of getting the params in the `/token` endpoint just by adding the fix. That's impossible since `/auth` is requested from the user agent whereas `/token` is requests from the OIDC RP.

I was just trying to find out whether in the code flow (as this is the only 2 leg flow) `extraParams` should apply in both endpoints. Consider the following flow as an example:

```
UA                RP                          OIDC
/sso_signin?foo=1
------------------>
                  | store "foo" in local session
301 to OIDC /auth?foo=1
<------------------
/auth?foo=1
------------------------------------------------->
                  Return the code
                  <-------------------------------
                  | fetch "foo" from session
                  POST /token?foo=1
                  ------------------------------->
                  | Return ID token
                  <-------------------------------
Redirect to callback URL
<------------------
```

Long story short, `extraParams` will always apply only to authorization endpoints.

@panva Thanks a lot for your prompt and detailed replies.

**Comment by panva** [comment]
`ctx.oidc.body` is not a hack, it's just that i didn't want to polute the `ctx` with the usual `ctx.request.body`. If you need to access the parsed body oidc-provider is working with, that's where you'll find it.

`extraParams` as documented enriches the authorization and interaction since there is a need to whitelist and pass only what's needed to interactions through multiple requests. That, plus the authorization context can be easily manipulated and adding hundreds of params could polute your adapter storages, etc.
